### DISCUSSION TITLE
Java SDK for Deepgram

### DISCUSSION THREAD

**Discussion by shehrooz-sattar** [deepgram/community#674] [Feature Requests]
Requesting Java SDK feature for deepgram.

Is there any Plan to introduce Java SDK for Deepgram?
if there is, then can you please provide details about the timeline of this feature?

**Comment by team-deepgram** [comment]
Thanks for asking your question about Deepgram! If you didn't already include it in your post, please be sure to add as much detail as possible so we can assist you efficiently, such as:
- The `request_id` if you have a question about your requests or transcription responses.
- The features you used or the full `api.deepgram.com` URL you sent your request to, including parameters.
- Any code snippets you can share.

**Comment by SandraRodgers** [comment]
We do not have a Java SDK at this time. We have talked about creating one, but there isn't an official plan yet. 

If you are interested in creating a community-run Java-sdk, that is an option! We could possibly work with you to support that.

**Reply by shehrooz-sattar** [reply]
This could be an exciting opportunity, and I would like to discuss more about it.

**Reply by jpvajda** [reply]
@shehrooz-sattar 👋  if you'd like to join us on our Discord Server we can engage with you further if you are considering developing a Community SDK for Java.  Just let us know your level of interest!


https://dpgr.am/discord

**Comment by shehrooz-sattar** [comment]
@SandraRodgers 

Thank you for your reply. 

Can you please confirm that everything that we can do through a Python SDK, we can also achieve using the Deepgram APIs or is there any difference between SDK functionality and APIs?

I know that we can use a pre-signed URL to upload the transcription files directly to the S3 bucket using the SDKs. Is the same possible by using deepgram APIs directly, as there is no SDK for Java? If possible, could you please provide an example?

Or is the Deepgram API going to provide the transcript only, and we have to take care of the uploading to the S3 bucket ourselves using a pre-signed URL?

**Reply by davidvonthenen** [reply]
There isn't any difference between the SDK and API barring any new features that come in.

There has been a lot of talk about a Java SDK... just waiting on a few things. Will keep you posted.

**Reply by SandraRodgers** [reply]
We have this documentation that gives information about using Presigned URLs: https://developers.deepgram.com/docs/using-aws-s3-presigned-urls-with-the-deepgram-api

I believe this documentation uses the Python SDK, however, as the example. But like my colleague said, you can do the same thing with the API as with the SDK.

The documentations says:

Presigned URLs that upload files to S3 can use either the PUT or POST HTTP method. It is critical that the PUT HTTP method is used when communicating with Deepgram. Presigned POST URLs move the credential information from the URL's query parameters into the request body, and since a callback's request body cannot be set via the Deepgram API the PUT method is required.

So I believe you can make a PUT request to the API to do what you are asking.

**Reply by shehrooz-sattar** [reply]
@SandraRodgers 

Thank you for your reply. Thank you for explaining the "put" method. I have seen the same already in the documentation and understand it.

Yes, we have been trying to achieve the same using the direct API for some time now without any success.


Our API call looks like the following...

curl \
  --request POST \
  --header 'Authorization: Token <API Token>' \
  --header 'Content-Type: application/json' \
  --data '{"url":"<get object s3 presigned url>","callback":"<put object s3 presigned url>","callback_method":"put"}' \
  --url 'https://api.deepgram.com/v1/listen?model=nova-2&smart_format=true'

  So we are including a "put object s3 pre-signed URL" in the callback to save the transcript directly to s3 and we are setting the callback_methods to "put".

As best of my understanding of the documentation, this should work but actually, it's not working and returns the transcript results in response to this API call synchronously rather than saving in the s3.

We have tried experimenting withh STT Callbacks in the API call but no success as I believe this is to use your custom webhooks rather than the S3 destination. https://developers.deepgram.com/docs/callback

curl \
  --request POST \
  --header 'Authorization: Token <API Token>' \
  --header 'Content-Type: application/json' \
  --data '{"url":"<get object s3 presigned url>"}' \
  --url 'https://api.deepgram.com/v1/listen?model=nova-2&smart_format=true&callback_method=put&callback=<put object s3 presigned url>'

This makes an asynchronous call but fails with 400 error for "s3 pre-signed url". However, the same url works fine in python sdk.
  
We have tried many configurations and gone through documentation a couple of times but no success. The article about pre-signed urls do not provide any insights on how to achieve using the direct api : https://developers.deepgram.com/docs/using-aws-s3-presigned-urls-with-the-deepgram-api#uploading-transcripts-directly-to-s3

it would be great if we have an example of the API call to achieve the same behaviour.

**Comment by shehrooz-sattar** [comment]
I was able to figure it out. The documentation is not clear about it and doesn't not provide any clear example that might  be improved over time. For example in many places,  Documentation says callback_method needs to be true or false rather than actual method name  - put or post.

Following is the way to make it work with Direct API Call.

curl \
  --request POST \
  --header 'Authorization: Token < API Key >' \
  --header 'Content-Type: application/json' \
  --data '{"url":" < presigned url for get Object in S3 >"}' \
  --url 'https://api.deepgram.com/v1/listen?model=nova-2&smart_format=true&callback_method=put&callback=<Url Encoded - presigned url for Put Object in S3>'
